test/e2e/periodic_test.go

```diff
@@ -45,5 +52,94 @@ var _ = Describe("ControlPlaneMachineSet Operator", framework.Periodic(), func()
 			})
 		})
 
+		Context("and an instance is terminated on the cloud provider", func() {
+			var machine *machinev1beta1.Machine
+			var idx int
+
+			BeforeEach(func() {
+				client := testFramework.GetClient()
+				machineList := &machinev1beta1.MachineList{}
+				machineSelector := runtimeclient.MatchingLabels(framework.ControlPlaneMachineSetSelectorLabels())
+
+				By("Getting a list of all control plane machines")
+				Expect(client.List(testFramework.GetContext(), machineList, machineSelector)).To(Succeed(), "should be able to retrieve list of control plane machines")
+
+				machine = &machineList.Items[0]
+
+				By("Checking that the machine is actually in index 0")
+				machineIdx, err := helpers.MachineIndex(*machine)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(machineIdx).To(Equal(0))
+
+				idx = machineIdx
+
+				By("Deleting an instance from the cloud provider")
+				Expect(testFramework.DeleteAnInstanceFromCloudProvider(machine)).To(Succeed())
+
+				By("Waiting for the machine to get into Failed phase")
+				Eventually(komega.Object(machine), 15*time.Minute).Should(HaveField("Status.Phase", HaveValue(Equal("Failed"))))
+
+				By("Deleting a control plane machine in phase Failed at index 0")
+				Expect(client.Delete(testFramework.GetContext(), machine)).To(Succeed())
+			})
+
+			helpers.ItShouldReplaceTheOutDatedMachineInDeleting(testFramework, idx)
+		})
+
+		Context("and a node with terminated kubelet", func() {
+			var client runtimeclient.Client
+			var ctx context.Context
+			var delObjects map[string]runtimeclient.Object
+			var machine *machinev1beta1.Machine
+			var idx int
+
+			BeforeEach(func() {
+				client = testFramework.GetClient()
+				ctx = testFramework.GetContext()
+				delObjects = make(map[string]runtimeclient.Object)
+				machineList := &machinev1beta1.MachineList{}
+				machineSelector := runtimeclient.MatchingLabels(framework.ControlPlaneMachineSetSelectorLabels())
+				node := &corev1.Node{}
+
+				By("Getting a list of all control plane machines")
+				Expect(client.List(ctx, machineList, machineSelector)).To(Succeed(), "should be able to retrieve list of control plane machines")
+
+				machine = &machineList.Items[0]
+
+				By("Checking that the machine is actually in index 0")
+				machineIdx, err := helpers.MachineIndex(*machine)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(machineIdx).To(Equal(0))
```

Comment:
Please add a note here to explain the expectation
```suggestion
				Expect(machineIdx).To(Equal(0), "Expected first machine to have index 0")
```
I expect this is fragile and is why the tests are failing right now, I think you may want a better way to extract the index 0 machine from this list and in the `BeforeEach`

Reply:
I suspect that the instance type increase test that runs before this is non-deterministic (we don't know what machine gets updated first time-wise).

Do we even need to check the idx? We always get the machine at 0 from the machineList. I don't really get what machine index this actually is (we get a correct idx from the `helpers.MachineIndex` to plug into `	helpers.ItShouldReplaceTheOutDatedMachineInDeleting` later). What functionality are we even testing by this check?

Reply:
> We always get the machine at 0 from the machineList

I don't think that's correct looking at the periodic-pre. I think this is failing because you've got index 2 from the list rather than index 0. I suspect, because there are changes to the cluster, that index 0 isn't alphabetically first always, and so this is causing a flake.

You need to make sure you're working within Index 0 here as this is the one you've told to be deleted, so, if for any reason the list order is not with the index 0 first, you'll fail this test later in weird ways.

This is definitely an important check IMO

Reply:
Came up with something. I will see how CI behaves.

Loop through the control plane machines until we find the expected index and return this machine and work with it.